posts/2019-10-1-serverless-azure-functions-v1-part2.md

```diff
@@ -0,0 +1,337 @@
+npm test
+# Serverless not contained within dev dependencies to avoid conflicts
+# because most users have it installed globally on their dev machine
+npm i serverless -g
+# Deploy service
+sls deploy
+```
+
+There are a lot more bells and whistles we could add, but that's essentially what it boils down to. Of course, we'll need authentication in whatever system we're deploying from, and that's where the [service principal](https://github.com/serverless/serverless-azure-functions#creating-a-service-principal) will come in. I'll show you how to use the service principal in the `deploy.yml` pipeline below.
+
+For my pipelines, I'm actually going to split up my CI and CD into `unit-tests.yml` and `deploy.yml`. Unit tests will be run on PRs into `master` or `dev` (this is assuming there are branch policies in place to prevent devs from pushing straight to either branch). Deployment will be run on commits (merges) to `master`.
+
+##### Unit Tests
+```yaml
+# pipelines/unit-tests.yml
+
+# Only run on Pull Requests into `master` or `dev`
+pr:
+  branches:
+    include:
+    - master
+    - dev
+
+# Run pipeline on node 8 and 10 on Linux, Mac and Windows 
+strategy:
+  matrix:
+    Linux_Node8:
+      imageName: 'ubuntu-16.04'
+      node_version: 8.x
+    Linux_Node10:
+      imageName: 'ubuntu-16.04'
+      node_version: 10.x
+    Mac_Node8:
+      imageName: 'macos-10.14'
+      node_version: 8.x
+    Mac_Node10:
+      imageName: 'macos-10.14'
+      node_version: 10.x
+    Windows_Node8:
+      imageName: 'win1803'
+      node_version: 8.x
+    Windows_Node10:
+      imageName: 'win1803'
+      node_version: 10.x
+
+# https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops#use-a-microsoft-hosted-agent
+pool:
+  vmImage: $(imageName)
+
+steps:
+- task: NodeTool@0
+  inputs:
+    versionSpec: $(node_version)
+  displayName: 'Install Node.js'
+
+# Make pipeline fail if tests or linting fail, linting occurs in `pretest` script
+- bash: |
+    set -euo pipefail
+    npm ci
+    npm test
+  displayName: 'Run tests'
+```
+
+##### Deployment
+```yaml
+# pipelines/deploy.yml
+
+trigger:
+  branches:
+    include:
+    - master
+
+# https://docs.microsoft.com/en-us/azure/devops/pipelines/library/variable-groups?view=azure-devops&tabs=yaml
+variables:
+- group: sls-deploy-creds
+
+jobs:
+
+- job: "Deploy_Azure_Function_App"
+  timeoutInMinutes: 30
+  cancelTimeoutInMinutes: 1
+
+  pool:
+    vmImage: 'ubuntu-16.04'
+
+  steps:
+  - task: NodeTool@0
+    inputs:
+      versionSpec: 10.x
+    displayName: 'Install Node.js'
+
+  - bash: |
+      npm install -g serverless
+    displayName: 'Install Serverless'
+    # Deploy service with prefix `gh`, stage `prod` and to region `West Europe`
+  - bash: |
+      npm ci
+      sls deploy --prefix gh --stage prod --region "West Europe"
+    env:
+      # Azure Service Principal. Secrets need to be mapped here
+      AZURE_SUBSCRIPTION_ID: $(AZURE_SUBSCRIPTION_ID)
```

Comment:
Are the service principle credentials supposed to be pasted into that config here? Or are they set somewhere else as an environment variable in whatever system runs this deploy.yml? This part wasn't clear to me but maybe it's fine to have it clarified in the documentation you link to.